### ISSUE TITLE
Proposal for adding functionality 'calculating + accessing individual feature contributions in the case that link="logit"'

### ISSUE THREAD

**Issue by jasper-vdm** [shap/shap#1365]
In some cases (e.g. when applying the TreeExplainer to a LightGBM classifier for the UCI Adult income dataset), it does not seem currently possible to directly extract the expected value and the SHAP values in a 'link="identity"'-like way.

When instantiating an explainer with link="identity", the base_value[i] and model_output_value[i][j] for output class i and instance j from test dataset X_test are simply equal to

- base_value[i] = explainer.expected_value[i]
- model_output_value[i][j] = explainer.expected_value[i] + sum(explainer.shap_values(X=X_test)[i][j,:] )

This has the advantage that not only the base_value[i] and model_output_value[i][j] are straightforward to calculate, but also that the each individual feature contribution contribution_feature_k[i][j,k] for the k-th feature is directly equal to its corresponding SHAP value, i.e.

- contribution_feature_k[i][j,k] =  explainer.shap_values(X=X_test)[i][j,k]

On the other hand, if we are restricted to using link="logit" (e.g. when applying the TreeExplainer to a LightGBM classifier for the UCI Adult income dataset), we have to calculate base_value[i] and model_output_value[i][j,:] for class i and instance j from test dataset X_test as follows:

- base_value[i] = convert_func(explainer.expected_value[i])
- model_output_value[i][j] = convert_func(explainer.expected_value[i] + sum(explainer.shap_values(X=X_test)[i][j,:]))

whereby convert_func = lambda x: 1 / (1 + np.exp(-x)) is the _logit_inverse-function (also defined in shap/shap/links.py). This approach is used in the function format_data(data) in shap/shap/plots/_force_matplotlib.py to prepare the data for plotting. 

Calculating the individual feature contributions (contribution_feature_k[i][j,k] for the k-th feature) to the model_output_value[i][j,:] is a more complex challenge and seems ambiguous since its contribution depends on when the contribution is added to x in the _logit_inverse-function 
e.g. imagine that all SHAP values explainer.shap_values(X=X_test)[i][j,:] for output class i and instance j of X_test have the same sign, the contribution contribution_feature_k[i][j,k] of the k-th feature to the model_output_value[i][j] will be larger when calculated using the following equation

- contribution_feature_k[i][j,k] = convert_func(explainer.expected_value[i] + explainer.expected_value(X=X_test)[i][j,k]) -  convert_func(explainer.expected_value[i])

than when calculated using the following equation

- contribution_feature_k[i][j,k] = convert_func(explainer.expected_value[i] + sum(explainer.shap_values(X=X_test)[i][j,:])) - convert_func(explainer.expected_value[i] + sum(explainer.shap_values(X=X_test)[i][j,:]) - explainer.expected_value(X=X_test)[i][j,k])

The correct way to calculate the feature contribution contribution_feature_k[i][j,k] is then to treat the different possible sums in a Shapley-like way. However, this is computationally very inefficient. Therefore, I would like to suggest to use a stepwise approximate method, thereby ensuring that the following equation remains valid:

- model_output_value[i][j] = base_value[i] + sum(contribution_feature_k[i][j,:])

I have experimented a little bit with this method and it gives promising results. Would it be worthwile adding a functionality to the shap-package so that these approximate contributions contribution_feature_k[i][j,:] as well as base_value[i] and model_output_value[i][j] can be accessed directly by the user in the case that link="logit"?
